test/unit/org/apache/cassandra/cql3/validation/entities/TupleTypeTest.java

```diff
@@ -233,5 +258,136 @@ public void testReversedTypeTuple() throws Throwable
+            SortedMap<ByteBuffer, Integer> map = new TreeMap<>(Comparator.comparing(currentTableMetadata().partitioner::decorateKey));
+            int count = 0;
+            for (ByteBuffer value : testcase.uniqueRows)
+            {
+                map.put(value, count);
+                ByteBuffer[] tupleBuffers = tupleType.split(value);
+
+                execute("INSERT INTO %s (id, value) VALUES (?, ?)", tuple(tupleBuffers), count);
+
+                assertRows(execute("SELECT * FROM %s WHERE id = ?", tuple(tupleBuffers)),
+                           row(tuple(tupleBuffers), count));
+                count++;
+            }
+            assertRows(execute("SELECT * FROM %s LIMIT 100"),
+                       map.entrySet().stream().map(e -> row(e.getKey(), e.getValue())).toArray(Object[][]::new));
+        }));
+    }
+
+    @Test
+    public void tupleCkReadWriteAsc()
+    {
+        tupleCkReadWrite(Order.ASC);
+    }
+
+    @Test
+    public void tupleCkReadWriteDesc()
+    {
+        tupleCkReadWrite(Order.Desc);
+    }
+
+    private void tupleCkReadWrite(Order order)
+    {
+        qt().withExamples(1).withShrinkCycles(0).forAll(caseGen()).checkAssert(orFail(testcase -> {
+            TupleType tupleType = testcase.type;
+            createTable("CREATE TABLE %s (pk int, ck " + toCqlType(tupleType) + ", value int, PRIMARY KEY(pk, ck))" +
+                        " WITH CLUSTERING ORDER BY (ck "+order.name()+")");
+            String cql = SchemaCQLHelper.getTableMetadataAsCQL(currentTableMetadata(), false, false, false);
+            SortedMap<ByteBuffer, Integer> map = new TreeMap<>(order.apply(tupleType::compare));
+            int count = 0;
+            for (ByteBuffer value : testcase.uniqueRows)
+            {
+                map.put(value, count);
+                ByteBuffer[] tupleBuffers = tupleType.split(value);
+
+                execute("INSERT INTO %s (pk, ck, value) VALUES (?, ?, ?)", 1, tuple(tupleBuffers), count);
+
+                assertRows(execute("SELECT * FROM %s WHERE pk = ? AND ck = ?", 1, tuple(tupleBuffers)),
+                           row(1, tuple(tupleBuffers), count));
+                count++;
+            }
+            UntypedResultSet results = execute("SELECT * FROM %s LIMIT 100");
+            List<Object[]> rows = toObjects(results);
+            Assertions.assertThat(rows.stream().mapToInt(o -> (int) o[2]).toArray())
+                      .as("value equality\n%s", cql)
+                      .isEqualTo(map.entrySet().stream().mapToInt(e -> e.getValue()).toArray());
+            assertRows(results,
+                       map.entrySet().stream().map(e -> row(1, e.getKey(), e.getValue())).toArray(Object[][]::new));
+        }));
+    }
+
+    private static final class Case {
+        TupleType type;
+        List<ByteBuffer> uniqueRows;
+    }
+
+    private static Gen<Case> caseGen()
+    {
+        int numRows = 10;
```

Comment:
nit: Might be nice to have `caseGen()` take `numRows` as an argument, even if all it does is document a little closer to the caller that we're using 10 rows. (It could conceivably inform the LIMIT as well.)